File: my-folder/2343-count-unguarded-cells-in-the-grid/solution.cpp
class Solution {
public:
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        
        // Gaurd = 2
        // empty = -1
        // wall = 3
        vector<vector<int>> matrix(m, vector<int>(n,-1));

        for (auto& g:guards) matrix[g[0]][g[1]] = 2;
        for (auto& w:walls) matrix[w[0]][w[1]] = 3;
        
        vector<vector<int>> directions = {{1,0}, {-1,0}, {0,1}, {0,-1}}; 

        for (auto& guard:guards)
        {
            // go in each direction as far as possible
            int x = guard[0];
            int y = guard[1];  
            for (auto& dir: directions)
            {
                int new_x = x + dir[0];
                int new_y = y + dir[1];
                while (new_x>=0 && new_x<m && new_y>=0 && new_y<n && matrix[new_x][new_y]!=3 && matrix[new_x][new_y]!=2)
                {
                    matrix[new_x][new_y] = 0;
                    new_x = new_x + dir[0];
                    new_y = new_y + dir[1];
                }
            }
        }

        int ans = 0;
        for (int i= 0; i<m; i++)
        {
            for (int j=0; j<n; j++)
            {
                if (matrix[i][j] == -1) ans+=1;
            }
        }
         return ans;



    }
};

File: my-folder/2343-count-unguarded-cells-in-the-grid/solution.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        GUARD, WALL, CLEAR = 3, 2, -1
        grid = [[CLEAR]*n for _ in range(m)]
        for gi,gy in guards: grid[gi][gy] = GUARD
        for wi,wy in walls: grid[wi][wy] = WALL
        directions = [[1,0],[-1,0],[0,1],[0,-1]]
        for gx,gy in guards:
            for dire in directions:
                new_x , new_y = gx + dire[0], gy + dire[1]
                while new_x >=0 and new_x < m and new_y >=0 and new_y < n and grid[new_x][new_y]!= GUARD and grid[new_x][new_y]!=WALL:
                    grid[new_x][new_y] = 0
                    new_x , new_y = new_x + dire[0], new_y + dire[1]
        
        return sum( [1 for j in range(n) for i in range(m) if grid[i][j] == CLEAR ])


            
            
        

